New Resource aws_redshift_integration

p2sk-ann opened the pull request:
### Description

This PR is to add a new resource aws_redshift_integration to the Redshift Family.

### Relations

Closes #40618.
Closes #34713.

### References

 - https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/RedshiftforDynamoDB-zero-etl.html
 - https://docs.aws.amazon.com/redshift/latest/dg/loading-data-copy-job.html
 - https://docs.aws.amazon.com/ja_jp/redshift/latest/APIReference/API_CreateIntegration.html
 - https://docs.aws.amazon.com/ja_jp/redshift/latest/APIReference/API_ModifyIntegration.html
 - https://docs.aws.amazon.com/ja_jp/redshift/latest/APIReference/API_DescribeIntegrations.html
 - https://docs.aws.amazon.com/ja_jp/redshift/latest/APIReference/API_DeleteIntegration.html

### Output from Acceptance Testing

```console
$ make testacc TESTS=TestAccRedshiftIntegration PKG=redshift
make: Verifying source code with gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go1.23.7 test ./internal/service/redshift/... -v -count 1 -parallel 20 -run='TestAccRedshiftIntegration'  -timeout 360m -vet=off
2025/04/03 14:00:28 Initializing Terraform AWS Provider...
=== RUN   TestAccRedshiftIntegration_basic
=== PAUSE TestAccRedshiftIntegration_basic
=== RUN   TestAccRedshiftIntegration_disappears
=== PAUSE TestAccRedshiftIntegration_disappears
=== RUN   TestAccRedshiftIntegration_optional
=== PAUSE TestAccRedshiftIntegration_optional
=== RUN   TestAccRedshiftIntegration_sourceUsesS3Bucket
=== PAUSE TestAccRedshiftIntegration_sourceUsesS3Bucket
=== CONT  TestAccRedshiftIntegration_basic
=== CONT  TestAccRedshiftIntegration_optional
=== CONT  TestAccRedshiftIntegration_disappears
=== CONT  TestAccRedshiftIntegration_sourceUsesS3Bucket
--- PASS: TestAccRedshiftIntegration_basic (286.39s)
--- PASS: TestAccRedshiftIntegration_optional (286.56s)
--- PASS: TestAccRedshiftIntegration_disappears (295.98s)
--- PASS: TestAccRedshiftIntegration_sourceUsesS3Bucket (368.38s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/redshift   374.858s
```

masaki-hirose-kinto commented:
This is my first Pull Request submission to this repository.
I have carefully read the contribution guide, run the CI locally, and made the necessary code modifications.
All tests pass with the following command:

```
make ci-quick
```

Additionally, I made one small local change to the GNUmakefile to successfully run 

```
make ci
```

Specifically, in the website-tflint section, I removed the following line:

```
"--disable-rule=aws_worklink_website_certificate_authority_association_invalid_certificate" \
```

This line caused TFLint to fail locally with the error: "Failed to check rule config; Rule not found".
Removing the line resolved the issue and allowed all checks to pass.
Since this issue might be specific to my local environment, I did not commit the change to the GNUmakefile.

I have also implemented a sweeper and confirmed it works with the following command:

```
SWEEPARGS=-sweep-run=aws_redshift_integration make sweep
```

masaki-hirose-kinto commented on internal/service/redshift/integration.go:
```diff
@@ -0,0 +1,390 @@
+	return "If source_arn is an S3 ARN and ConfigValue is null, do not show any differences."
+}
+
+func (m *ignoreKmsKeyIdForS3Modifier) PlanModifyString(ctx context.Context, req planmodifier.StringRequest, resp *planmodifier.StringResponse) {
```
The implementation of this planmodifier is based on the following background:

The CreateIntegration() API allows specifying either a DynamoDB table or an S3 bucket as the SourceARN. However, the parameters kms_key_id and additional_encryption_context are only valid for DynamoDB sources. If you include these parameters when using an S3 bucket, the API returns the error:

> The KMS KeyId and Additional Encryption Context parameters are only valid for a zero-ETL integration with Amazon DynamoDB sources.

For a DynamoDB-based integration, if kms_key_id is not provided, an Amazon-owned KMS Key is automatically used. Therefore, kms_key_id must be marked optional and computed in our code. On the other hand, if you specify an S3 bucket as the SourceARN, the CreateIntegration() API response does not include a kms_key_id. Below is an example CloudTrail log showing the response from CreateIntegration() when the SourceARN is an S3 bucket:

```
"responseElements": {
        "integrationArn": "arn:aws:redshift:us-west-2:xxxxxx:integration:ff151402-2884-4ba7-bdcc-ce83dec6d5e1",
        "integrationName": "s3-integration-hoge",
        "sourceArn": "arn:aws:s3:::sample-s3-redshift-integration",
        "sourceType": "s3-event-notifications",
        "targetArn": "arn:aws:redshift-serverless:us-west-2:xxxxxx:namespace/c2f4bc36-309b-48f2-9f7a-2092bf391df7",
        "status": "creating",
        "errors": [],
        "createTime": "Mar 31, 2025, 4:21:42 AM",
        "description": "test",
        "tags": [
            {
                "key": "foo",
                "value": "bar"
            }
        ]
    },
```

Because of this, when CreateIntegration() is called with an S3 bucket, kms_key_id is never set in the Terraform state (i.e., terraform state show aws_redshift_integration.test does not display it). As a result, if you run terraform plan to change a property normally updated through ModifyIntegration() (such as the integration_name), Terraform interprets kms_key_id as “(known after apply) # forces replacement,” triggering a full resource recreation.

To avoid that behavior, I implemented a planmodifier so that any changes to kms_key_id go undetected when the SourceARN is an S3 bucket. However, if the user explicitly specifies kms_key_id in the configuration, I decided it is more appropriate to return the API error directly, and therefore excluded those cases from the planmodifier’s condition.

I am not entirely certain this implementation is the perfect approach for all scenarios, but based on a variety of tests (beyond the acceptance tests), I have found no major issues with it in practice.

Additionally, because additional_encryption_context is not set as Computed (i.e., Computed = false), we concluded that a similar implementation is not necessary for it.

p2sk-ann commented on internal/service/redshift/integration.go:
```diff
@@ -0,0 +1,390 @@
+	return "If source_arn is an S3 ARN and ConfigValue is null, do not show any differences."
+}
+
+func (m *ignoreKmsKeyIdForS3Modifier) PlanModifyString(ctx context.Context, req planmodifier.StringRequest, resp *planmodifier.StringResponse) {
```
The implementation of this planmodifier is based on the following background:

The CreateIntegration() API allows specifying either a DynamoDB table or an S3 bucket as the SourceARN. However, the parameters kms_key_id and additional_encryption_context are only valid for DynamoDB sources. If you include these parameters when using an S3 bucket, the API returns the error:

> The KMS KeyId and Additional Encryption Context parameters are only valid for a zero-ETL integration with Amazon DynamoDB sources.

For a DynamoDB-based integration, if kms_key_id is not provided, an Amazon-owned KMS Key is automatically used. Therefore, kms_key_id must be marked optional and computed in our code. On the other hand, if you specify an S3 bucket as the SourceARN, the CreateIntegration() API response does not include a kms_key_id. Below is an example CloudTrail log showing the response from CreateIntegration() when the SourceARN is an S3 bucket:

```
"responseElements": {
        "integrationArn": "arn:aws:redshift:us-west-2:xxxxxx:integration:ff151402-2884-4ba7-bdcc-ce83dec6d5e1",
        "integrationName": "s3-integration-hoge",
        "sourceArn": "arn:aws:s3:::sample-s3-redshift-integration",
        "sourceType": "s3-event-notifications",
        "targetArn": "arn:aws:redshift-serverless:us-west-2:xxxxxx:namespace/c2f4bc36-309b-48f2-9f7a-2092bf391df7",
        "status": "creating",
        "errors": [],
        "createTime": "Mar 31, 2025, 4:21:42 AM",
        "description": "test",
        "tags": [
            {
                "key": "foo",
                "value": "bar"
            }
        ]
    },
```

Because of this, when CreateIntegration() is called with an S3 bucket, kms_key_id is never set in the Terraform state (i.e., terraform state show aws_redshift_integration.test does not display it). As a result, if you run terraform plan to change a property normally updated through ModifyIntegration() (such as the integration_name), Terraform interprets kms_key_id as “(known after apply) # forces replacement,” triggering a full resource recreation.

To avoid that behavior, I implemented a planmodifier so that any changes to kms_key_id go undetected when the SourceARN is an S3 bucket. However, if the user explicitly specifies kms_key_id in the configuration, I decided it is more appropriate to return the API error directly, and therefore excluded those cases from the planmodifier’s condition.

I am not entirely certain this implementation is the perfect approach for all scenarios, but based on a variety of tests (beyond the acceptance tests), I have found no major issues with it in practice.

Additionally, because additional_encryption_context is not set as Computed (i.e., Computed = false), we concluded that a similar implementation is not necessary for it.

ewbankkit replied:
This can be accomplished with the [`RequiresReplaceIfConfigured`](https://developer.hashicorp.com/terraform/plugin/framework/resources/plan-modification#requiresreplaceifconfigured) plan modifier.

p2sk-ann commented:
This is my first Pull Request submission to this repository.
I have carefully read the contribution guide, run the CI locally, and made the necessary code modifications.
All tests pass with the following command:

```
make ci-quick
```

Additionally, I made one small local change to the GNUmakefile to successfully run 

```
make ci
```

Specifically, in the website-tflint section, I removed the following line:

```
"--disable-rule=aws_worklink_website_certificate_authority_association_invalid_certificate" \
```

This line caused TFLint to fail locally with the error: "Failed to check rule config; Rule not found".
Removing the line resolved the issue and allowed all checks to pass.
Since this issue might be specific to my local environment, I did not commit the change to the GNUmakefile.

I have also implemented a sweeper and confirmed it works with the following command:

```
SWEEPARGS=-sweep-run=aws_redshift_integration make sweep
```

ewbankkit reviewed (approved):
LGTM :rocket:.

```console
% make testacc TESTARGS='-run=TestAccRedshiftIntegration_' PKG=redshift ACCTEST_PARALLELISM=2
make: Verifying source code with gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go1.23.8 test ./internal/service/redshift/... -v -count 1 -parallel 2  -run=TestAccRedshiftIntegration_ -timeout 360m -vet=off
2025/04/13 17:32:33 Initializing Terraform AWS Provider...
=== RUN   TestAccRedshiftIntegration_basic
=== PAUSE TestAccRedshiftIntegration_basic
=== RUN   TestAccRedshiftIntegration_disappears
=== PAUSE TestAccRedshiftIntegration_disappears
=== RUN   TestAccRedshiftIntegration_optional
=== PAUSE TestAccRedshiftIntegration_optional
=== RUN   TestAccRedshiftIntegration_sourceUsesS3Bucket
=== PAUSE TestAccRedshiftIntegration_sourceUsesS3Bucket
=== RUN   TestAccRedshiftIntegration_tags
=== PAUSE TestAccRedshiftIntegration_tags
=== CONT  TestAccRedshiftIntegration_basic
=== CONT  TestAccRedshiftIntegration_sourceUsesS3Bucket
--- PASS: TestAccRedshiftIntegration_sourceUsesS3Bucket (245.34s)
=== CONT  TestAccRedshiftIntegration_optional
--- PASS: TestAccRedshiftIntegration_basic (256.45s)
=== CONT  TestAccRedshiftIntegration_tags
--- PASS: TestAccRedshiftIntegration_optional (295.05s)
=== CONT  TestAccRedshiftIntegration_disappears
--- PASS: TestAccRedshiftIntegration_tags (310.28s)
--- PASS: TestAccRedshiftIntegration_disappears (340.82s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/redshift	887.134s
```

ewbankkit commented:
```console
% make golangci-lint PKG=redshift
make: golangci-lint Checks / 1 of 5...
make: golangci-lint Checks / 2 of 5...
make: golangci-lint Checks / 3 of 5...
make: golangci-lint Checks / 4 of 5...
make: golangci-lint Checks / 5 of 5...
```

jar-b reviewed (commented):
One test suggestion, otherwise LGTM

ewbankkit commented on internal/service/redshift/integration_test.go:
```diff
@@ -0,0 +1,742 @@
+					testAccCheckIntegrationExists(ctx, resourceName, &integration),
+					acctest.CheckFrameworkResourceDisappears(ctx, acctest.Provider, tfredshift.ResourceIntegration, resourceName),
+				),
+				ExpectNonEmptyPlan: true,
```
Done.

ewbankkit replied:
```console
% make testacc TESTARGS='-run=TestAccRedshiftIntegration_disappears' PKG=redshift
make: Verifying source code with gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go1.23.8 test ./internal/service/redshift/... -v -count 1 -parallel 20  -run=TestAccRedshiftIntegration_disappears -timeout 360m -vet=off
2025/04/14 10:41:16 Initializing Terraform AWS Provider...
=== RUN   TestAccRedshiftIntegration_disappears
=== PAUSE TestAccRedshiftIntegration_disappears
=== CONT  TestAccRedshiftIntegration_disappears
--- PASS: TestAccRedshiftIntegration_disappears (281.26s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/redshift	286.854s
```

jar-b reviewed (approved):
LGTM 🚀

masaki-hirose-kinto commented:
@ewbankkit Thanks for the kind words! I’m happy to contribute.

p2sk-ann commented:
@ewbankkit Thanks for the kind words! I’m happy to contribute.
